### ISSUE TITLE
H264/VCE: Different VBV Buffer Sizes cause inconsistent encoding effects (Permanent&Temporary Quality Loss, Encoder Failure, etc)

### ISSUE THREAD

**Issue by Xaymar** [GPUOpen-LibrariesAndSDKs/AMF#53]
The VCE based H264 encoder currently produces varying broken results depending on the VBV Buffer Size. This is especially noticeable on Complex/Moving to Simple/Still scene changes. Some of the effects include:

- *Permanent Quality Loss*: Also known as 'I-Frame Refreshing', though it is actually the inverse of that. Each I-Frames QP values are so terrible that it is beyond hope to get anything useable from those frames. P-Frames then pick up on the problem and recover the quality. Except that the next I-Frame has the same quality as the previous one and this never stops.
- *Temporary Quality Loss*: The normal reaction, if it always were somewhat the same length. Length of this effect can vary between several frames to several seconds, depending on the VBV Buffer Size.
- *Missing Target Bitrate*: Another problem, the encoder can miss the Target Bitrate so much that it ends up breaking any connection that exists. Example would be 3500kbit CBR and it would use 200-400kbit in one second and in the next go up into the 10-20mbit range and then slowly go back to the Target Bitrate. Except at that point it is too late, the network connection is saturated and frames are being dropped - possibly forever.
- *Filler Data failure*: I don't actually know why this happens, Filler Data should not be affected by this at all.
- ~~*Bluescreens*: One of the more rare ones, some VBV Buffer Sizes actually cause a Bluescreen. Unknown DLL and address, so I believe this to be caused by hardware doing things it shouldn't.~~
- ~~*Temporary Encoder failure*: Hard to explain, lasts until cutting power from the system. Rebooting will just freeze Windows at the boot up screen. Attempting to use the encoder in a session where this happened will freeze Windows or cause a Bluescreen.~~

This primarily affects Constant Bitrate but Variable Bitrate is also broken due to this.

## How to reproduce this issue:
**Reproduction Toolkit:** http://cdn.xaymar.com/private/2017/01/09/03/AMF53.7z (extract and run either portable32.bat or portable 64.bat to start it)

1. Set the Encoder to the GPU you want to test, with or without OpenCL (has no effect on output).
2. Toy around with Minimum QP and VBV Buffer Size. I personally only tested Min QP 18, 11, 6 and 0 and VBV Buffer Size at really strict (Bitrate/FPS), loose (Bitrate) and unrestricted (100000).
3. Mute any audio source in the OBS mixer if they aren't muted.
4. Toggle the Media Source quickly and then hit Start Recording.
5. Record for ~30 seconds
6. Analyze the output using Bitrate Viewer, ffprobe or any other tools.

## Examples of this happening:
All Examples use Constant Bitrate at a Target Bitrate of 3500 and a Min QP of 0. Left image is average over a second, right is per frame bitrate. The dotted grey line is the average over a longer period of time, the blue part is below/at the average bitrate, the purple/red part above and the black horizontal line delimits 5000kbit.

### [VBV Buffer at maximum possible Size (100000)](http://cdn.xaymar.com/private/2017/01/11/19/2017-01-11%2019-41-52.mkv)
![Avg Per Second](http://cdn.xaymar.com/private/2017/01/11/19/BitrateViewer_2017-01-11_19-43-26.png) ![Frame Bitrate](http://cdn.xaymar.com/private/2017/01/11/19/BitrateViewer_2017-01-11_19-43-35.png)

### [VBV Buffer at 16x Target Bitrate](http://cdn.xaymar.com/private/2017/01/11/19/2017-01-11%2019-39-47.mkv)
![Avg Per Second](http://cdn.xaymar.com/private/2017/01/11/19/BitrateViewer_2017-01-11_19-41-26.png) ![Frame Bitrate](http://cdn.xaymar.com/private/2017/01/11/19/BitrateViewer_2017-01-11_19-41-35.png)

### [VBV Buffer at 8x Target Bitrate](http://cdn.xaymar.com/private/2017/01/11/19/2017-01-11%2019-38-01.mkv)
![Avg Per Second](http://cdn.xaymar.com/private/2017/01/11/19/BitrateViewer_2017-01-11_19-39-01.png) ![Frame Bitrate](http://cdn.xaymar.com/private/2017/01/11/19/BitrateViewer_2017-01-11_19-39-09.png)

### [VBV Buffer at 4x Target Bitrate](http://cdn.xaymar.com/private/2017/01/11/19/2017-01-11%2019-33-19.mkv)
![Avg Per Second](http://cdn.xaymar.com/private/2017/01/11/19/BitrateViewer_2017-01-11_19-36-01.png) ![Frame Bitrate](http://cdn.xaymar.com/private/2017/01/11/19/BitrateViewer_2017-01-11_19-37-36.png)

### [VBV Buffer at 2x Target Bitrate](http://cdn.xaymar.com/private/2017/01/11/19/2017-01-11%2019-31-57.mkv)
![Avg Per Second](http://cdn.xaymar.com/private/2017/01/11/19/BitrateViewer_2017-01-11_19-32-49.png) ![Frame Bitrate](http://cdn.xaymar.com/private/2017/01/11/19/BitrateViewer_2017-01-11_19-33-03.png)

### [VBV Buffer at 1x Target Bitrate](http://cdn.xaymar.com/private/2017/01/11/19/2017-01-11%2019-21-08.mkv)
![Avg Per Second](http://cdn.xaymar.com/private/2017/01/11/19/BitrateViewer_2017-01-11_19-26-39.png) ![Frame Bitrate](http://cdn.xaymar.com/private/2017/01/11/19/BitrateViewer_2017-01-11_19-26-27.png)

### [VBV Buffer at 1/2 Target Bitrate](http://cdn.xaymar.com/private/2017/01/11/19/2017-01-11%2019-25-23.mkv)
![Avg Per Second](http://cdn.xaymar.com/private/2017/01/11/19/BitrateViewer_2017-01-11_19-26-51.png) ![Frame Bitrate](http://cdn.xaymar.com/private/2017/01/11/19/BitrateViewer_2017-01-11_19-27-04.png)

### [VBV Buffer at 1/4 Target Bitrate](http://cdn.xaymar.com/private/2017/01/11/19/2017-01-11%2019-27-50.mkv)
![Avg Per Second](http://cdn.xaymar.com/private/2017/01/11/19/BitrateViewer_2017-01-11_19-29-31.png) ![Frame Bitrate](http://cdn.xaymar.com/private/2017/01/11/19/explorer_2017-01-11_19-29-42.png)

### [VBV Buffer at 1/30 Target Bitrate](http://cdn.xaymar.com/private/2017/01/11/23/2017-01-11%2023-10-18.mkv)
![Avg Per Second](http://cdn.xaymar.com/private/2017/01/11/23/BitrateViewer_2017-01-11_23-11-47.png) ![Frame Bitrate](http://cdn.xaymar.com/private/2017/01/11/23/explorer_2017-01-11_23-11-57.png)

### [VBV Buffer at 1/60 Target Bitrate](http://cdn.xaymar.com/private/2017/01/11/23/2017-01-11%2023-08-25.mkv)
![Avg Per Second](http://cdn.xaymar.com/private/2017/01/11/23/BitrateViewer_2017-01-11_23-11-24.png) ![Frame Bitrate](http://cdn.xaymar.com/private/2017/01/11/23/BitrateViewer_2017-01-11_23-11-34.png)

### [VBV Buffer at "1.1 * target_bitrate / frame_rate"](http://cdn.xaymar.com/private/2017/01/11/23/2017-01-11%2023-12-52.mkv)
![Avg Per Second](http://cdn.xaymar.com/private/2017/01/11/23/BitrateViewer_2017-01-11_23-13-41.png) ![Frame Bitrate](http://cdn.xaymar.com/private/2017/01/11/23/BitrateViewer_2017-01-11_23-14-03.png)

-----

This has been reported to me here:

* https://github.com/Xaymar/obs-studio_amf-encoder-plugin/issues/180
* https://github.com/Xaymar/obs-studio_amf-encoder-plugin/issues/183

I shortly afterwards reproduced it, analyzed it, tried to figure out what happens and the results are the above. Edit: First reported here: https://github.com/GPUOpen-LibrariesAndSDKs/AMF/issues/40

**Comment by Xaymar** [comment]
I have prepared a test setup as always: http://cdn.xaymar.com/private/2017/01/09/03/AMF53.7z .

Repro-Steps:

1. Set the Encoder to the GPU you want to test, with or without OpenCL (has no effect on output).
2. Toy around with Minimum QP and VBV Buffer Size. I personally only tested Min QP 18, 11, 6 and 0 and VBV Buffer Size at really strict (Bitrate/FPS), loose (Bitrate) and unrestricted (100000).
3. Mute any audio source in the OBS mixer if they aren't muted.
4. Toggle the Media Source quickly and then hit Start Recording.
5. Record for ~30 seconds
6. Analyze the output using Bitrate Viewer, ffprobe or any other tools.

It is a 1 in 4 chance that it happens. You will be immediately able to see it in the output file if it happened.

**Comment by Xaymar** [comment]
The effects of VBV Buffer Size aren't exact either. At some point they start doing the opposite effect and some actually work out properly. Here's with which VBV Buffer Sizes it happens most often:

- VBV Buffer Size=100000: Always - Lasts forever, extreme effect, hard to explain exact effect on average bitrate
- VBV Buffer Size=Target Bitrate x16: Lasts for about five seconds, average bitrate drops to 0
- VBV Buffer Size=Target Bitrate x8: Lasts for about two seconds, average bitrate drops to 0
- VBV Buffer Size=Target Bitrate x4: Lasts for a second, average bitrate drops to almost 0
- VBV Buffer Size=Target Bitrate x2: Lasts for half a second, average bitrate drops by 1500kbit
- VBV Buffer Size=Target Bitrate: Minimal impact, average bitrate drops by 800kbit before recovering
- VBV Buffer Size=Target Bitrate /2: High Impact, average bitrate swaps between extremely high and extremely low once it starts happening. This is the I-Frame Refreshing Effect.
- VBV Buffer Size=Target Bitrate /4: High Impact, average bitrate swaps between extremely high and extremely low once it starts happening. This is the I-Frame Refreshing Effect.


Target Bitrate was 3500kbit here, Min QP 0. The only one where it didn't happen is when the VBV Buffer Size is equal to the Target Bitrate, but that has it's own side effects again. Increasing Min QP to 18 makes it less pronounced, but lowers overall quality

**Comment by SuslikV** [comment]
@Xaymar Is this about the reverse "keyframe pumping"?
That was mentioned there: https://obsproject.com/forum/threads/comparison-of-x264-nvenc-quicksync-vce.57358/page-3#post-251548

**Comment by Xaymar** [comment]
@SuslikV No.

**Comment by SuslikV** [comment]
@Xaymar Then you need to post screens of the bitrate viewer where each frame is visible (scale up timeline), I think.

**Comment by Xaymar** [comment]
@SuslikV I will wait for maintainers to reply and ask for more information. If you want to try to reproduce this yourself, see this reply: https://github.com/GPUOpen-LibrariesAndSDKs/AMF/issues/53#issuecomment-271202237

**Comment by SuslikV** [comment]
@Xaymar Just few images to complete the opinion:
AMD (frames)
![amd frames clipboard-5](https://cloud.githubusercontent.com/assets/19683044/21857498/5487f418-d82e-11e6-8f2a-caa1d128f80a.png)
NV (frames)
![nv frames clipboard-6](https://cloud.githubusercontent.com/assets/19683044/21857511/61d6bf3c-d82e-11e6-81c5-9ef3bcb0b736.png)
Software (frames)
![software frames clipboard-4](https://cloud.githubusercontent.com/assets/19683044/21857522/6dadb6da-d82e-11e6-8f10-cb7563371143.png)
AMD (seconds)
![amd seconds clipboard-8](https://cloud.githubusercontent.com/assets/19683044/21857547/80fac1e2-d82e-11e6-8e69-49908688edeb.png)
NV (seconds)
![nv seconds clipboard-7](https://cloud.githubusercontent.com/assets/19683044/21857561/89a5e74a-d82e-11e6-85e9-2b34ed546d7d.png)
Software (seconds)
![software seconds clipboard-9](https://cloud.githubusercontent.com/assets/19683044/21857567/8dbe6348-d82e-11e6-947f-9b4ea1db01d3.png)
This is almost the same fragment of the video sample (from the link above). Averaging is not clear thing when you cannot control the time window and scale. That's all I wanted to say.

**Comment by Xaymar** [comment]
@SuslikV 
While it's great that you participate in this issue, the content of your replies has so far not been relevant to the actual issue at all. If you actually want to contribute anything, please use the repro-kit I linked you to instead of the comparison files that are literally ancient in terms of computer age.

Additionally "average bitrate" is only ever measured over a sequence (1 second) or the entire file - in fact with Bitrate Viewer you can choose how it measures it. And for streaming or constant bitrate, a average bitrate window of 1 second is the only relevant thing. Anyway, here are the images you wanted, not that they are in any way useful:

[VBV at maximum](http://cdn.xaymar.com/private/2017/01/11/19/2017-01-11%2019-41-52.mkv)
![second based](http://cdn.xaymar.com/private/2017/01/11/19/BitrateViewer_2017-01-11_19-43-26.png)
![frame based](http://cdn.xaymar.com/private/2017/01/11/19/BitrateViewer_2017-01-11_19-43-35.png)

[VBV at sixteen times Target Bitrate](http://cdn.xaymar.com/private/2017/01/11/19/2017-01-11%2019-39-47.mkv)
![second based](http://cdn.xaymar.com/private/2017/01/11/19/BitrateViewer_2017-01-11_19-41-26.png)
![frame based](http://cdn.xaymar.com/private/2017/01/11/19/BitrateViewer_2017-01-11_19-41-35.png)

[VBV at octuple Target Bitrate](http://cdn.xaymar.com/private/2017/01/11/19/2017-01-11%2019-38-01.mkv)
![second based](http://cdn.xaymar.com/private/2017/01/11/19/BitrateViewer_2017-01-11_19-39-01.png)
![frame based](http://cdn.xaymar.com/private/2017/01/11/19/BitrateViewer_2017-01-11_19-39-09.png)

[VBV at quadruple Target Bitrate](http://cdn.xaymar.com/private/2017/01/11/19/2017-01-11%2019-33-19.mkv)
![second based](http://cdn.xaymar.com/private/2017/01/11/19/BitrateViewer_2017-01-11_19-36-01.png)
![frame based](http://cdn.xaymar.com/private/2017/01/11/19/BitrateViewer_2017-01-11_19-37-36.png)

[VBV at double Target Bitrate](http://cdn.xaymar.com/private/2017/01/11/19/2017-01-11%2019-31-57.mkv)
![second based](http://cdn.xaymar.com/private/2017/01/11/19/BitrateViewer_2017-01-11_19-32-49.png)
![frame based](http://cdn.xaymar.com/private/2017/01/11/19/BitrateViewer_2017-01-11_19-33-03.png)

[VBV at Target Bitrate](http://cdn.xaymar.com/private/2017/01/11/19/2017-01-11%2019-21-08.mkv)
![second based](http://cdn.xaymar.com/private/2017/01/11/19/BitrateViewer_2017-01-11_19-26-39.png)
![frame based](http://cdn.xaymar.com/private/2017/01/11/19/BitrateViewer_2017-01-11_19-26-27.png)

[VBV at half Target Bitrate](http://cdn.xaymar.com/private/2017/01/11/19/2017-01-11%2019-25-23.mkv)
![second based](http://cdn.xaymar.com/private/2017/01/11/19/BitrateViewer_2017-01-11_19-26-51.png)
![frame based](http://cdn.xaymar.com/private/2017/01/11/19/BitrateViewer_2017-01-11_19-27-04.png)

[VBV at quarter Target Bitrate](http://cdn.xaymar.com/private/2017/01/11/19/2017-01-11%2019-27-50.mkv)
![second based](http://cdn.xaymar.com/private/2017/01/11/19/BitrateViewer_2017-01-11_19-29-31.png)
![frame based](http://cdn.xaymar.com/private/2017/01/11/19/explorer_2017-01-11_19-29-42.png)

Please also note that apparently this happens less with A's video converter when reducing the GOP Size.

**Comment by acp2** [comment]
This also happens with Relive on my 260x & 270 , can provide recordings if needed.

**Comment by SuslikV** [comment]
@Xaymar It doesn't matter how ancient my examples are. Each frame is present on the my screens. I have shown only 365 frames (~12 sec) of the AMD encoded video, but yours - it is averaged data again. I simply saying that you need give exact data to show what is wrong. Build **the base** to compare to. Finally, post the screen how it should look (if this possible). But exact screen, not averaging data. That's all.

_More:_
About screens. 

- Even in my screens I don't know what the vertical scale is (I think it is constant value of size of "something").

About method. 

- If it were a quality issue (visual) I would ask you for the footage screenshot - original, actual, expected (if possible).
- You must know what you want to be fixed, and way it must to be fixed.
- I don't like Bitrate Viewer.

**Comment by Xaymar** [comment]
@SuslikV I'm not sure what's worse, your inability to read or your inability to understand. Let me break this down for you as simple as possible (aka "the dumbing down process"):

> It doesn't matter how ancient my examples are.

Yes it does. Each version of the plugin and driver improved things and the examples you used were from driver 16.9.2 - there is at least one firmware update between that version and the current version 16.12.2. I've created those files, I know what's in them and what settings I used to make them. 

> Each frame is present on the my screens. I have shown only 365 frames (~12 sec) of the AMD encoded video, but yours - it is averaged data again

What? This was never about frames, it was about quality instability which users referred to as I-Frame Refreshing. Also my screenshots are one averaged per second and one that shows individual frames - not averages. Please actually read and understand before you reply.

> I simply saying that you need give exact data to show what is wrong. Build the base to compare to. Finally, post the screen how it should look (if this possible). But exact screen, not averaging data. That's all.

Have you even taken a look at the reproduction kit I linked you to? It's all in there!

> Even in my screens I don't know what the vertical scale is (I think it is constant value of size of "something").

Take a look at the left side, it says what each increment of the black line means.

> If it were a quality issue (visual) I would ask you for the footage screenshot - original, actual, expected (if possible).

Again, see the reproduction kit and the files i linked in the comment you replied to.

> You must know what you want to be fixed, and way it must to be fixed.

Did you even read anything in here? It's right there in the original issue post. Also I can't know how AMF/VCE works internally, as it is a closed source application.

> I don't like Bitrate Viewer.

I don't care. Use ffprobe then, it will show you the exact same issue. Case closed.

And finally, this german saying fits perfectly here:

> Herr, wirf Hirn vom Himmel!
> Oder Steine, hauptsache er trifft!

**Comment by MikhailAMD** [comment]
Hi, here is the comment from rate control developer: 
I checked their description and settings in “recordEncoder.json”. They are using large VBV buffer (3500) for CBR mode and target bitrate is 3500.  
Larger vbv buffer -> more room to absorb complexity change -> more time to converge to target bitrate after “complex -> simple” scene change.
If they are expecting quick response, we suggest to set vbv buffer size to smaller value e.g. “1.1 * target_bitrate / frame_rate”.

**Comment by Xaymar** [comment]
I've updated the original report to include more information and also put all the examples in there including VBV Buffer at 1/30th, 1/60th and 1.1*TargetBitrate/FrameRate.

**Comment by SuslikV** [comment]
@Xaymar At least, you can post a video encoded with software x264 (if it anyway better in sense of bitrate padding).

**Comment by MikhailAMD** [comment]
Thanks for the description update. It is very informative. After analyzing the description and the attached files rate control developer said that you must always enable filler data for CBR or behavior will be unpredictable - it is disabled right now in the reproduction toolkit.

**Comment by Xaymar** [comment]
I think there was some misunderstanding again, so I'll try to clear things up: the problem is not the average bitrate not being at the Target Bitrate. The true problem is that even with Filler Data enabled, AMDs encoder produces Frame-Flickering, as you can see in the following files:

### [VBV Size at 100000](http://cdn.xaymar.com/private/2017/01/14/01/vce_3500_VBV100000.mkv)
![Average Bitrate](http://cdn.xaymar.com/private/2017/01/14/01/BitrateViewer_2017-01-14_01-15-22.png) ![Bitrate per Frame](http://cdn.xaymar.com/private/2017/01/14/01/explorer_2017-01-14_01-15-37.png)
This file is by far the worst, it didn't even use any of the settings that were given to it.

### [VBV Size at 16x Bitrate](http://cdn.xaymar.com/private/2017/01/14/01/vce_3500_VBV056000.mkv)
![Average Bitrate](http://cdn.xaymar.com/private/2017/01/14/01/BitrateViewer_2017-01-14_01-16-19.png) ![Bitrate per Frame](http://cdn.xaymar.com/private/2017/01/14/01/BitrateViewer_2017-01-14_01-15-53.png)
Here we can see that it tries to fix the extreme Bitrate variation, but it reacts several seconds later instead of in the next frame. Lowest Bitrate is 400kbps which slowly returns to 3500kbps, only to then overshoot it and then undershoot it again before returning to normal. If you view the file itself, you can see exactly one I-Frame Refresh effect happening.

### [VBV Size at 8x Bitrate](http://cdn.xaymar.com/private/2017/01/14/01/vce_3500_VBV028000.mkv)
![Average Bitrate](http://cdn.xaymar.com/private/2017/01/14/01/explorer_2017-01-14_01-17-56.png) ![Bitrate per Frame](http://cdn.xaymar.com/private/2017/01/14/01/BitrateViewer_2017-01-14_01-18-04.png)
Similar to the above, but the duration and extremes are not as bad. After the switch from motion to static there are two I-Frame Refreshing effects, one stronger than the other.

### [VBV Size at 4x Bitrate](http://cdn.xaymar.com/private/2017/01/14/01/vce_3500_VBV014000.mkv)
![Average Bitrate](http://cdn.xaymar.com/private/2017/01/14/01/BitrateViewer_2017-01-14_01-18-42.png) ![Bitrate per Frame](http://cdn.xaymar.com/private/2017/01/14/01/BitrateViewer_2017-01-14_01-19-12.png)
Again similar to the above but weaker. I-Frame Refreshing Effect takes places over multiple frames but is still longer than it should be.

### [VBV Size at 2x Bitrate](http://cdn.xaymar.com/private/2017/01/14/01/vce_3500_VBV007000.mkv)
![Average Bitrate](http://cdn.xaymar.com/private/2017/01/14/01/BitrateViewer_2017-01-14_01-19-39.png) ![Bitrate per Frame](http://cdn.xaymar.com/private/2017/01/14/01/explorer_2017-01-14_01-19-48.png)
Getting closer to the desired result, but not quite there yet. It takes approximately .5 seconds for it to recover the quality.

### [VBV Size at 1x Bitrate](http://cdn.xaymar.com/private/2017/01/14/01/vce_3500_VBV003500.mkv)
![Average Bitrate](http://cdn.xaymar.com/private/2017/01/14/01/BitrateViewer_2017-01-14_01-20-58.png) ![Bitrate per Frame](http://cdn.xaymar.com/private/2017/01/14/01/BitrateViewer_2017-01-14_01-21-05.png)
Almost perfect, this is what it should do: Immediately return to a good quality.

### [VBV Size at 1/2 Bitrate](http://cdn.xaymar.com/private/2017/01/14/01/vce_3500_VBV001750.mkv)
![Average Bitrate](http://cdn.xaymar.com/private/2017/01/14/01/BitrateViewer_2017-01-14_01-22-19.png), ![Bitrate per Frame](http://cdn.xaymar.com/private/2017/01/14/01/BitrateViewer_2017-01-14_01-22-34.png)
And here it starts getting worse again. While the bitrate drop gets smaller and smaller, the quality recover time gets longer and longer again.

### [VBV Size at 1/4 Bitrate](http://cdn.xaymar.com/private/2017/01/14/01/vce_3500_VBV000875.mkv)
![Average Bitrate](http://cdn.xaymar.com/private/2017/01/14/01/BitrateViewer_2017-01-14_01-23-53.png) ![Bitrate per Frame](http://cdn.xaymar.com/private/2017/01/14/01/BitrateViewer_2017-01-14_01-24-11.png)
See above.

### [VBV Size at 1/30 Bitrate](http://cdn.xaymar.com/private/2017/01/14/01/vce_3500_VBV000117.mkv)
![Average Bitrate](http://cdn.xaymar.com/private/2017/01/14/01/BitrateViewer_2017-01-14_01-24-33.png) ![Bitrate per Frame](http://cdn.xaymar.com/private/2017/01/14/01/BitrateViewer_2017-01-14_01-25-02.png)
Average Bitrate is almost perfect, if only the quality also were perfect. This one now has the extreme I-Frame Refreshing effect, as detailed in the first post already.

### [VBV Size at 1/60 Bitrate](http://cdn.xaymar.com/private/2017/01/14/01/vce_3500_VBV000058.mkv)
![Average Bitrate](http://cdn.xaymar.com/private/2017/01/14/01/BitrateViewer_2017-01-14_01-25-22.png) ![Bitrate per Frame](http://cdn.xaymar.com/private/2017/01/14/01/BitrateViewer_2017-01-14_01-25-39.png)
This one just... ignored Filler Data for some reason? It has less quality issues than the one above, but the average bitrate went to shit.

### [VBV Size at (Bitrate * 1.1 / FPS)](http://cdn.xaymar.com/private/2017/01/14/01/vce_3500_VBV000064.mkv)
![Average Bitrate](http://cdn.xaymar.com/private/2017/01/14/01/BitrateViewer_2017-01-14_01-26-14.png) ![Bitrate per Frame](http://cdn.xaymar.com/private/2017/01/14/01/BitrateViewer_2017-01-14_01-26-27.png)
Perfect average bitrate, not so perfect result in quality. See the 1/30 description as to what's wrong with the quality.

I have little knowledge of how the VCE firmware works, but I believe that the issue is due to a average bitrate window that is too large. As far as I know, with x264 the bitrate window is equal to the GOP size or the buffer size, since x264 also has variable GOP size support and keyframe on demand features.
As for VCE, I do not know what the actual bitrate window is. Since I can't control the GOP Size directly anymore (permanently stuck at 1/2 IDR Period, GOPSize and GOPAlignment both do nothing), it is not possible to fix this on my end unless the problem is caused by some other code.

Here are some x264 files for comparison: [Ultra Fast](http://cdn.xaymar.com/private/2017/01/14/02/x264_3500_ultrafast.mkv), [Super Fast](http://cdn.xaymar.com/private/2017/01/14/02/x264_3500_superfast.mkv), [Very Fast](http://cdn.xaymar.com/private/2017/01/14/02/x264_3500_veryfast.mkv), [Faster](http://cdn.xaymar.com/private/2017/01/14/02/x264_3500_faster.mkv), [Fast](http://cdn.xaymar.com/private/2017/01/14/02/x264_3500_fast.mkv).

**Comment by SuslikV** [comment]
@Xaymar just post screenshots from the both videos (x264 vs VCE) right after scene change (and record by AMD VCE at 30 fps too!).

@Xaymar Is the reverse to "keyframe pumping" (I mentioned above) was resolved?

@Xaymar from your previous samples I think that the "window" is too strict. ^_^

**Comment by SuslikV** [comment]
I think, the issue mentioned by @Xaymar as "I-Frame Refreshing Effect" is the sharpness of the text that restores slowly after scene change, see example below (GIF image, ~313KB, 25fps, ~2 sec):
![ezgif_com-622531f91d](https://cloud.githubusercontent.com/assets/19683044/21993939/04e71794-dc26-11e6-814c-aac0809455fe.gif)
this test image produced by software x264 encoder at very low bitrate (CBR ~50 kbit/s) in OBS Studio v17.0.0, (larger fragment of the footage inside the zip file, mp4 video 6 sec: [2017-01-16 20-42-31.zip](https://github.com/GPUOpen-LibrariesAndSDKs/AMF/files/709113/2017-01-16.20-42-31.zip) )

I think, it is more visible on the gif right now. So, is this it? Or the issue you are complaining of, is completely different, @Xaymar ?

**Comment by Xaymar** [comment]
@SuslikV Take a look at the files I uploaded. There is a reason why I uploaded them.

**Comment by SuslikV** [comment]
@Xaymar I didn't see any significant error from the posted videos. That the encoder is capable of. I don't think that any changes to AMF can solve issues from the footage.

**Comment by Xaymar** [comment]
> I didn't see any significant error from the posted videos. That the encoder is capable of. I don't think that any changes to AMF can solve issues from the footage.

I'm not sure if you actually looked at the video files or just claimed to have done so. Watch the parts directly after the switch from moving to still footage and you will see the exact issue I and many others have encountered and are trying to get fixed.

**Comment by acp2** [comment]
Maybe you can see an error here - http://s.go.ro/upzkfemr

**Comment by SuslikV** [comment]
@acp2 is your example recorded with latest software (OBS Studio plugin, driver etc.) and using default settings (any preset made by @Xaymar ), isn't it?

Because, I see the IDR of awful quality inserted in each frame from your example (key interval you are using is 2 sec, framerate 60 fps), nothing was changed since it was mentioned in the OBS forum thread (see link [above](https://github.com/GPUOpen-LibrariesAndSDKs/AMF/issues/53#issuecomment-271882977)).
just compare (the screenshots of the posted "2016-12-25 04-04-04.mp4" video cropped to the left bottom corner):
previous frame 239
![frame_239_r](https://cloud.githubusercontent.com/assets/19683044/22024122/91a321fe-dcd1-11e6-824c-48340493760d.png)

next frame 240 is IDR
![frame_240_r](https://cloud.githubusercontent.com/assets/19683044/22024138/a00a803e-dcd1-11e6-9d1e-6d2618375efb.png)

animated:
![ezgif_com-frames-239-240](https://cloud.githubusercontent.com/assets/19683044/22024103/834316dc-dcd1-11e6-849f-96504746997c.gif)
The difference well noticeable on the left part of the screenshots (brown object), at center (purple background) and present through whole the video footage you provided each 2 seconds (detail loss in each IDR frame, no matter static scene or dynamic move).

The quality of the image restores in time (until next IDR inserted). And **visually** this is the same to the video example I posted (software x264 at very low bitrate - the way how it restores frame by frame).

I didn't mentioned about visible blocking of the image because nothing to compare too (x264, NVIDIA), maybe it stays at decent level.

AMF can't do magic - this is software layer to encoder. That's all. the Best thing that should be done, IMHO - is to remove the VBV from the user part at all and hide it at the driver's profiles for encoding tasks ("target device" preset or so). The encoder itself maybe the subject to change.., or you must accept that it wasn't designed to work at low bitrates with HD videos at 60 fps.

I see, that higher bitrates (far more than 3500kbit/s) are looking good for AMD VCE videos or you have other thoughts?

**Comment by acp2** [comment]
The same is happening on a 540p30fps file with 3500 bitrate , so please , tell me how that should happen just because it's "hardware encoding".
This is a bug , something that should NOT be happening.

**Comment by Xaymar** [comment]
@SuslikV At this point I'm no longer sure if you are trolling or if you are actually this incapable of thinking. All the information you need to understand the issue is in front of you, now actually use it instead of living in your own world.

I recommend you watch [this](http://cdn.xaymar.com/private/2017/01/14/01/vce_3500_VBV000064.mkv), [this](http://cdn.xaymar.com/private/2017/01/14/01/vce_3500_VBV000117.mkv) and [this](http://cdn.xaymar.com/private/2017/01/14/01/vce_3500_VBV100000.mkv) to understand what the informed people in here are talking about.

> the Best thing that should be done, IMHO - is to remove the VBV from the user part at all and hide it at the driver's profiles for encoding tasks ("target device" preset or so). The encoder itself maybe the subject to change.., or you must accept that it wasn't designed to work at low bitrates with HD videos at 60 fps.

This is a problem in the VCE firmware which needs to be fixed. It happens with any VBV Buffer Size, any framerate and any Bitrate. The only thing this "solution" would do is push more people to NVidia or Intel.

---

@MikhailAMD Has there been any new information on why the effects from VBV Buffer Size are so inconsistent? The example 1.1*Bitrate/FPS and 1/30 showcase one of the effects, 100000 showcases another, 1/60 showcases one where Filler Data just stopped doing anything but the quality was perfect and the remaining ones showcases a mix of problems.
.

**Comment by MikhailAMD** [comment]
Rate control developer and I reviewed the videos Xaymar attached.
**To summarize**:
VBV: 100000 - very large size - takes long time to get to recover from bad quality to normal
VBV 16x bitrate: no flickering recovers somewhat faster
VBV 8x bitrate: no flickering recovers faster than x16
VBV 4x bitrate: no flickering recovers faster than x8
VBV 1x bitrate: the best
VBV 1/2x bitrate: these is I-frame flickering due to small budget for the I-Frame
VBV 1/4x bitrate: simialr to 1/2
VBV Bitrate *1 .1 / FPS: I-frame flickering is more pronounced due to the very small budget to I-Frame
Larger VBV buffer could help to prevent “I frame refreshing effect” for static scene, because it leaves more bit budget for I-frame.
On the other hand, larger VBV buffer could take longer time to converge to target bitrate when “complex->simple” scene change happens. 
So as the results suggested, a good trade off could be set VBV buffer to 1x bitrate.

**In short**: Larger VBV buffer means large budget for any frame including I-frame but too large VBV means longer recovery.

**Comment by SuslikV** [comment]
room for I-frames restricted by level and frame size (3.0...5.2 similar to number of references), yeah?
